### PR TITLE
feat: integrate with `rai_state_logs`

### PR DESCRIPTION
**Pull request by boczekbartek**
For diff base of this PR is https://github.com/RobotecAI/rai/pull/324  -> merge after it. 

## Purpose

- `rai_state_logs` has been a separate module not used by [RaiStateBasedLlmNode](https://github.com/RobotecAI/rai/blob/1dd21d335c86070e8b8c69b2eb66222db0dcfef3/src/rai/rai/node.py#L338) and `rai` [examples](https://github.com/RobotecAI/rai/tree/development/examples) 
- logs were collected to a buffer and summarized by a simple llm model.
- llm inferance on ros logs can be costly and time consuming. 

## Proposed Changes

- [x] abstract logs parsing for state-based agents by creating a simple interface
- [x] implement rai_state_logs parser using the interface
- [x] refactor existing llm summarizer to use the interface
- [x] parser type (`rai_state_logs` or `llm`) can be set as `RaiStateBasedLlmNode` parameter.
- [x] use `rai_state_logs` as a default logs parser
- [x] add `rai_state_logs` to rai examples 
- [x] hacky handling of empty `filters` list was added due to: https://github.com/ros2/rclcpp/issues/1955

Related PRs and issues that should be resolved before merging this PR:
- https://github.com/RobotecAI/rai/pull/322
- because `rai_state_logs` will be on by default after thie PR is merged, if this issue is not resoled, examples will not work: https://github.com/RobotecAI/rai/issues/323
    - https://github.com/RobotecAI/rai/pull/324

## Issues

- https://github.com/RobotecAI/rai/issues/170

## Testing

- How was it tested, what were the results?

## Summary by CodeRabbit

- **New Features**
	- Introduced a new logging parser system for enhanced log management.
	- Added functionality for summarizing logs from ROS nodes.
	- Integrated state logging functionality into the launch process.
	- Enhanced command-line argument handling for dynamic allowlist configuration.
	- Added new launch configurations for TurtleBot4 and Rai Whoami.
	- Expanded the allowlist with various ROS topics for improved operational capabilities.

- **Bug Fixes**
	- Implemented a check to clear empty filters in the log digest node constructor.

- **Documentation**
	- Updated launch argument defaults for improved clarity in log filtering.

- **Refactor**
	- Restructured logging and state management processes for better modularity and flexibility.
	- Removed outdated logging mechanisms to streamline the codebase.

### CODE REVIEW

**Comment by boczekbartek** [comment]
@coderabbitai full review

**Comment by boczekbartek** [comment]
@coderabbitai full review

File: src/rai/rai/node.py [hunk 1]
```diff
@@ -400,8 +393,18 @@ def __init__(
             state_retriever=self.get_robot_state,
             logger=self.get_logger(),
         )
+
+        # We have to use a separate node that we can manually spin for ros-service based
+        # parser and this node ros-subscriber based parser
+        logs_parser_node = self if logs_parser_type == "llm" else self._async_tool_node
+        self.logs_parser = create_logs_parser(
+            logs_parser_type, logs_parser_node, callback_group=self.callback_group
+        )
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply src/rai/rai/node.py:402]: its validated in create_logs_parser

File: src/rai/rai/utils/ros_logs.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    callback_group: Optional[rclpy.callback_groups.ReentrantCallbackGroup] = None,
+    bufsize: Optional[int] = 100,
+) -> BaseLogsParser:
+    if parser_type == "rai_state_logs":
+        return RaiStateLogsParser(node)
+    elif parser_type == "llm":
+        if any([v is None for v in [llm, callback_group, bufsize]]):
+            raise ValueError(
+                "Must provide llm, callback_group, qos_profile, and bufsize"
+            )
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply src/rai/rai/utils/ros_logs.py]: good catch

**Review by adamdbrw — APPROVED** [review]

File: src/rai/rai/utils/ros_logs.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        self.append(f"[{msg.stamp.sec}][{msg.name}]:{msg.msg}")
+
+    def clear(self):
+        self._buffer.clear()
+
+    def append(self, line: str):
+        self._buffer.append(line)
+        if len(self._buffer) > self.bufsize:
+            self._buffer.popleft()
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply src/rai/rai/utils/ros_logs.py:144]: https://docs.python.org/3/library/collections.html
>  Deques support thread-safe, memory efficient appends and pops from either side of the deque with approximately the same O(1) performance in either direction.

File: src/rai/rai/node.py [hunk 4]
```diff
@@ -346,6 +345,7 @@ def __init__(
         observe_postprocessors: Optional[Dict[str, Callable[[Any], Any]]] = None,
         allowlist: Optional[List[str]] = None,
         tools: Optional[List[Type[BaseTool]]] = None,
+        logs_parser_type: Literal["llm", "rai_state_logs"] = "rai_state_logs",
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply src/rai/rai/node.py:348]: Yes, we don't want to expose it yet

File: src/rai_bringup/launch/sim_whoami_demo.launch.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+from launch.launch_description_sources import PythonLaunchDescriptionSource
+from launch.substitutions import LaunchConfiguration
+from launch_ros.substitutions import FindPackageShare
+
+
+def generate_launch_description():
+    robot_description_package = LaunchConfiguration("robot_description_package")
+    game_launcher = LaunchConfiguration("game_launcher")
+    allowlist = LaunchConfiguration("allowlist")
+    demo_script = LaunchConfiguration("demo_script")
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply src/rai_bringup/launch/sim_whoami_demo.launch.py]: It's taken by imported launch - `rai_whoami_node`

File: src/rai_bringup/launch/sim_whoami_demo.launch.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            ),
+            DeclareLaunchArgument(
+                "game_launcher",
+                description="Path to the O3DE game launcher executable",
+            ),
+            DeclareLaunchArgument(
+                "allowlist",
+                default_value="src/examples/turtlebot4/allowlist.txt",
+                description="A list of ros interfaces that are exposed to rai agent.",
+            ),
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply src/rai_bringup/launch/sim_whoami_demo.launch.py]: `rai_examples` is not a package yet

File: examples/rosbot-xl.launch.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                    "demo_script": "examples/rosbot-xl-demo.py",
+                    "robot_description_package": "rosbot_xl_whoami",
+                    "game_launcher": game_launcher,
+                }.items(),
+            ),
+            ExecuteProcess(
+                cmd=["bash", "run-nav.bash"],
+                cwd="src/examples/rai-rosbot-xl-demo",
+                output="screen",
+            ),
```
[the remark this answers is not in the document]
Comment (boczekbartek) [reply examples/rosbot-xl.launch.py]: `rai_examples` is not a ros package

**Comment by boczekbartek** [comment]
I've just resolved conflicts and rebased on `development`

**Review by maciejmajek — APPROVED** [review]
Works as intended. ~Please add 
"Do not use cmd_vel topic to navigate.- use appropriate ros2 actions instead." to the demo's system prompt.
I just can't look at the poor LLM trying to navigate using cmd vel...~

Edit: This might not be the most suitable PR for that. Let's do it in the next one (update system prompts to enhance multiple demos at once)

**Pull request merged**